### DISCUSSION TITLE
--clientname does not change the jack client name on Windows 10

### DISCUSSION THREAD

**Discussion by newlaurent62** [jamulussoftware/jamulus#1678] [General]
Hello,

the --clientname option does not change the jack client name in qjackctl.

Example:
`Jamulus --clientname TEST
`
![Jamulus_clientname](https://user-images.githubusercontent.com/10225394/118025592-50471480-b360-11eb-82d3-33656802f888.PNG)

It works on Linux but does not work on Windows 10. 

The jack client name should be "Jamulus TEST".

Kind Regards,
Laurent

Tested on Jamulus 3.7.0 (Windows and Linux)

**Comment by hoffie** [comment]
Hi,

thanks for the report.
Just to confirm, this is on a custom Windows build with `CONFIG+="jackonwindows"`, right?

I currently do not see any platform-specific code in the jack name handling, so I'm a bit lost why this problem occurs.

**Comment by newlaurent62** [comment]
Yes the test has been done with the official 3.7.0 windows build and the Linux version is from Librazik 3 repository. 

I modified the jackrouter ini file. I will try to reinstall jack.

**Comment by newlaurent62** [comment]
I will firstly try to use another program that rename the jack client.

**Comment by newlaurent62** [comment]
So, i tried to reinstall Jamulus 3.7.0 (official build on github) and reinstall Jack v1.9.18 but jack client name does not change.

QJackTrip another software that i use works and i can set jack client name as i wish. 

I also tried Jamulus 3.6.2 but no luck too ...

Any idea ? Someone has the same behavior on Windows 10 (3.7.0 official build) ?

Please try in a terminal if you have intalled Jack : 
```
start "TEST" "c:\Program Files\Jamulus\Jamulus.exe" --server

start "TEST" "c:\Program Files\Jamulus\Jamulus.exe" --clientname TEST -c 127.0.0.1

```

The Jamulus client name option should : 
- set your jamulus windows title to "Jamulus TEST 127.0.0.1 - 1 User" (OK for me)
- set the jack client name to "Jamulus TEST" (Not OK for me).

Kind Regards,
Laurent

**Comment by hoffie** [comment]
> i tried to reinstall Jamulus 3.7.0 (official build on github)

Ah, so you are using the official Jamulus build. As far as I know, this is ASIO-only. If you want to use Jamulus directly with jack on Windows, you will have to do a custom build with `CONFIG+=jackonwindows`, as far as I understand.

You seem to be using `JackRouter` which is something completely different. It leads to a setup which looks like this: Jamulus (ASIO) -> JackRouter (Jack) -> Jackd. In this setup, Jamulus does not have a way to talk to jackd directly.

It may be possible to change the client name for ASIO as well and this may make it work, but this is a different feature request to what's already implemented.

**Comment by newlaurent62** [comment]
You are right : I'm using jack router asio driver.

If it's not possible to change jack client name with the asio driver then my demand is to support jackd directly on windows 10 too on the official build.

**Reply by hoffie** [reply]
> my demand is to support jackd directly on windows 10 too on the official build.

Jamulus already supports native jack on Windows. However, for this to work the whole audio layer in Jamulus has to be replaced, which cannot be done at runtime. Therefore, this is a build time option. The official Windows builds do not support jack, but rather the regular, direct ASIO interface which should cover the common use case.

Basically, I see three options to solve your case:
1. You can build your own Windows binaries with the special `CONFIG+=jackonwindows` option. With such a build, you would get native jack support and would not require JackRouter for Jamulus anymore. It should also make `--clientname` work.
2. We can think about providing additional Windows builds with this build option. However, we should weigh whether it is worth the effort and how to avoid confusing newcomers which do not need or want Jack (i.e. proper naming for such a build). Cc'ing @npostavs and @nefarius2001 who have worked on special Windows builds (Portaudio) and the build process in general respectively.
3. You can try to find out where the "Jamulus" label in JackRouter comes from. I initially suspected that Jamulus would provide that to the ASIO API and JackRouter would get it from there, but I haven't found such code (yet). So my next guess would have been the Window title, but then `--clientname` would have worked already. My last guess is that it's just the `.exe` name without extension, so you could give it a try to rename/copy `Jamulus.exe` to `Jamulus-ClientA.exe` and check if the name in Jack(Router) changes as well. If not, some further digging in JackRouter code may be needed.

PS: From the screenshots I assume that you are not a native English speaker (and neither am I). I just want to point out that "I demand" can be read to be rather harsh when talking about Open Source projects as everyone here is working on Jamulus in their free time. I'm pretty sure you rather meant something like "request" or "wish". :)

**Reply by newlaurent62** [reply]
> PS: From the screenshots I assume that you are not a native English speaker (and neither am I). I just want to point out that "I demand" can be read to be rather harsh when talking about Open Source projects as everyone here is working on Jamulus in their free time. I'm pretty sure you rather meant something like "request" or "wish". :)

You are right, my speaking language is french. Excuse me for my bad wording ...

**Reply by hoffie** [reply]
> You are right, my speaking language is french. Excuse me for my bad wording ...

No harm done, I just wanted to point it out ;)

**Reply by newlaurent62** [reply]
> My last guess is that it's just the .exe name without extension, so you could give it a try to rename/copy Jamulus.exe to Jamulus-ClientA.exe and check if the name in Jack(Router) changes as well

Yes, you're right. Changing the exe name changes the jack client name but i would rather prefer a jackonwindows official release :-).

**Reply by hoffie** [reply]
>  i would rather prefer a jackonwindows official release :-).

I understand. Personally, I lack PowerShell experience and test environments to add the necessary autobuild logic + jack installation (or even compilation). Let's see what the others say. :)

If this is realistic to add, a new issue for doing exactly that can be opened.

**Reply by henkdegroot** [reply]
> Yes, you're right. Changing the exe name changes the jack client name but i would rather prefer a jackonwindows official release :-).

When I compile on windows with "jackonwindows" the clientname option adds the name to the exe name. Something like Jamulus TEST is seen in JackControl. It seems to work with JACK and not JACK2 and only 64-bit. If you want I can share the installer I build....

**Reply by newlaurent62** [reply]
@henkdegroot Yes it would be nice. I will test it.

**Reply by newlaurent62** [reply]
Ideally the Jamulus for jack (Windows and MacOS) should have a different names and icons ... (May be add a jack to the Jamulus icon ?). Could it be possible that the two version jack and portaudio or coreaudio could coexist on the system ?

**Reply by henkdegroot** [reply]
@newlaurent62 , you can download from here to test: [jamulus-380-beta2-withjackonwindows](https://we.tl/t-BPkQkLhKUu)
The zip contains a windows-installer-exe but also the "program files" folder. Not sure what your preference is.

**Reply by newlaurent62** [reply]
@henkdegroot Sorry, but it does not work for me. Jamulus does not start. I think i have install jack v1.9.18 on windows too but if this is important i will check.

**Reply by henkdegroot** [reply]
@newlaurent62 , yes you need to have JACK installed as well and it has to be running as well. I have done a quick test with jack v1.9.11 64bit. If you have been running Jamulus on the same machine before, you may also need to start with a fresh ini file (not sure if this makes a difference) or remove the existing (default) jamulus.ini.

**Reply by newlaurent62** [reply]
@henkdegroot I tested your install package on a new Windows 10 VM with Jack 1.9.18 but same behavior. Jamulus does not start even if Jack server is started.

**Reply by newlaurent62** [reply]
I notice that the installer complains about missing ASIO driver on my fresh install. Note : For jack version ASIO driver is not needed because we can use Dummy driver with Jack Server.

**Reply by henkdegroot** [reply]
> I notice that the installer complains about missing ASIO driver on my fresh install. Note : For jack version ASIO driver is not needed because we can use Dummy driver with Jack Server.

The system I used for testing does have an ASIO driver installed. Looks like the Jamulus.pro file still includes references to ASIO (even when using the jackonwindows option).

**Reply by hoffie** [reply]
The ASIO warnings in installer and Jamulus have been touched recently, so maybe they do not handle the jackonwindows case correctly in all places. :(

**Reply by newlaurent62** [reply]
The only DAW i've found that support system devices and jack on all platforms is Ardour. It make me think that it's possible to support coreaudio and jack or portaudio and jack or alsa and jack.

**Reply by henkdegroot** [reply]
@newlaurent62 , I have just tested the installer on a "more or less" clean windows (at least no asio etc. and never had Jamulus installed). So I installed JACK and then ran the jamulus installer. I noticed the installer messages are just warnings and still allow to continue and install completes. Now starting Jamulus, I got a message to indicate jack server could not be found and auto start failed. I then start JACK manually and after this I was able to run Jamulus without problems.
I have tested on Win10 build 1909.

**Reply by newlaurent62** [reply]
This would be, in my opinion, a better way than creating two different builds.

**Reply by newlaurent62** [reply]
@henkdegroot yes i know it's just a warning. As i said i installed Jamulus, i have the icon but when i click the icon nothing happens.
![Jamulus-onWindows-2020](https://user-images.githubusercontent.com/10225394/118398706-1b85d680-b65a-11eb-9024-09e358780aa3.png)

**Reply by newlaurent62** [reply]
When i press enter i can input a command line ... Jamulus just stops without any apparent reason.

**Reply by newlaurent62** [reply]
My VM is the last Windows 10 version 20H2.

**Reply by newlaurent62** [reply]
I use jack v1.9.18 (last version).

**Reply by newlaurent62** [reply]
Thank you @henkdegroot to take time with me.

**Reply by henkdegroot** [reply]
I can only assume there is something else installed on my "more or less" clean machine. I am going to try on a really clean one and let you know.

**Reply by newlaurent62** [reply]
I did not install JackRouter ASIO driver.

**Reply by henkdegroot** [reply]
> I use jack v1.9.18 (last version).

Are you sure this is JACK and not JACK2? Currently seems the source code is prepared to work with JACK and not with JACK2.

**Reply by newlaurent62** [reply]
How can i setup JACK1 with jack V1.9.18 ? Is it possible ?

**Reply by henkdegroot** [reply]
No to my knowledge the latest JACK1 is v1.9.11. I believe some updates in the source is required to make things work with JACK2. I will try to compile a version with JACK2 and see if that generates errors.

**Reply by newlaurent62** [reply]
@henkdegroot There is something i don't understand. You say that jackonwindows uses JACK1, right ? On Linux i use jackd2 1.9.12 and on MacOS Mojave i use jack v1.9.18 with Jamulus.

**Reply by henkdegroot** [reply]
@newlaurent62 , I have just compiled with jack2 and that works. Just minor updates need in the Jamulus.pro file for this (as jack on windows has moved to a different folder). Guess Linux and MacOS use the same location or other logic to include this.
Here my screenshot:
![jamulus-with-jack2](https://user-images.githubusercontent.com/13550012/118402966-89d39480-b66c-11eb-9392-5b890a4eb3a8.png)

Url to download the [installer](https://we.tl/t-JXoQQOcEnY)

**Reply by newlaurent62** [reply]
Nice @henkdegroot, it works ! Thanks a lot !

Note : I have a button for asio devices in Jamulus parameters which is useless in this context.

I like the new GUI for Jamulus parameters.

**Reply by henkdegroot** [reply]
Guessing a bit more updates to the code would be needed to hide asio items when using only jack on Windows.

**Reply by newlaurent62** [reply]
Shall do i create an enhancement request for a jamulus build on Windows, MacOS and Linux allowing Jack server support and native sound support (Linux ALSA, Windows Portaudio and MacOS Coreaudio) ?

**Reply by henkdegroot** [reply]
I would say you can always propose the feauture/enhancement request....not sure if there are many users for it though.

**Reply by newlaurent62** [reply]
https://github.com/jamulussoftware/jamulus/discussions/1712

**Reply by jujudusud** [reply]
You have to install the latest JACK version for windows. and after that modify the Jamulus.pro file for the good paths (JACK2 instead of JACK)

**Reply by hoffie** [reply]
> Guessing a bit more updates to the code would be needed to hide asio items when using only jack on Windows.

Yep, I agree. Possibly also in the installer (which might be a bit harder?).

**Reply by henkdegroot** [reply]
> Possibly also in the installer (which might be a bit harder?).

As they are just warnings, I would leave the installer as it is. Seems a bit harder to check in the installer if you have compiled with jack or not.

**Comment by hoffie** [comment]
Note: I am converting this to a Github Discussion for now, as it is not clear yet if and what to do within Jamulus. We try to keep Issues for directly actionable bugs or enhancements. I'm going to reply in the Discussion thread which will be linked from here.
